### PR TITLE
Standardize the Dictionary interface, ensure the BYTES support

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
- Type conversion between INT/LONG/FLOAT/DOUBLE/STRING are allowed
- Type conversion between STRING/BYTES are allowed
- Push down and enhance the MutableDictionary range predicate evaluation
- Return dictId/dictIds while indexing value into MutableDictionary
- Add documentation for the contract of Dictionary interface
- Enhance PinotDataType for BYTES conversion

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java [hunk 1]
```diff
@@ -53,102 +53,148 @@ public BytesOffHeapMutableDictionary(int estimatedCardinality, int maxOverflowHa
...
   }
 
   @Override
-  public boolean inRange(@Nonnull String lower, @Nonnull String upper, int dictIdToCompare, boolean includeLower,
-      boolean includeUpper) {
-    return valueInRange(new ByteArray(BytesUtils.toBytes(lower)), new ByteArray(BytesUtils.toBytes(upper)),
-        includeLower, includeUpper, new ByteArray(getBytesValue(dictIdToCompare)));
+  public int indexOf(String stringValue) {
+    byte[] bytesValue = BytesUtils.toBytes(stringValue);
+    return getDictId(new ByteArray(bytesValue), bytesValue);
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java]: Here you can just use `BytesUtils. toByteArray(stringValue)` which is already in this PR. 😉

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java [hunk 2]
```diff
@@ -21,18 +21,40 @@
...
-   * @return Index of object, or -1 if not found.
+   * NOTE: Immutable dictionary is always sorted; mutable dictionary is always unsorted.
+   */
+  boolean isSorted();
+
+  int length();
+
+  /**
+   * Returns the index of the string representation of the value in the dictionary, or {@link #NULL_VALUE_INDEX} (-1) if
+   * the value does not exist. This API is for cross-type predicate evaluation.
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java]: Please add a comment here saying the stringValue is always unpadded.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/BytesUtils.java [hunk 3]
```diff
@@ -20,27 +20,38 @@
...
+
+  /**
+   * Converts a Hex encoded string to a {@link ByteArray}.
+   *
+   * @param stringValue Hex encoded string
+   * @return Decoded {@link ByteArray}
+   */
+  public static ByteArray toByteArray(String stringValue) {
+    try {
+      return new ByteArray(Hex.decodeHex(stringValue.toCharArray()));
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/BytesUtils.java]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java]: I need both byte[] and ByteArray, so I only fetched the byte[]

Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java]: Oh I see

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/Dictionary.java]: This value is from the query, so no concept of padding here. Added comments for the String binary search.

**Pull request merged**